app/scripts/controllers/metametrics.js

```diff
@@ -107,6 +139,108 @@ export default class MetaMetricsController {
     );
   }
 
```

Comment:
code below throws error in various scenario, I am wondering if that can cause issues for user

Reply:
@Gudahtt @kumavis what is the appropriate way of throwing errors for developers implementing events? @jpuri raises a good point and I'm just curious if anything has changed in my absence for best practices around throwing useful errors in background code that **should** be non user facing.

Reply:
This PR should have all of the relevant context: https://github.com/MetaMask/metamask-extension/pull/12573

Basically our strategy is the same as it always was - we should throw anything unexpected and ensure it's logged to Sentry (either explicitly or by not catching it). If it's expected but still shouldn't be user-facing, we can use console logs (and maybe a metric event, if it's something there is value in tracking).

The recent change in this PR was to make the `trackEvent` function non-blocking, for performance reasons. Typically there is no need to wait on submitting a metric event before proceeding to the next operation. The same approach might be useful here too?

There are actually now two functions for submitting events, one non-blocking (`trackEvent`), and one blocking (`submitEvent`), so that there would be an option in the rare case where we did need to block on an event getting submitted. For example, if we start using Mixpanel funnels, we might want to block on the first event getting submitted to ensure that it always gets sent first before the other funnel events (otherwise they might not be recognized by Mixpanel as being part of the same funnel). But as far as I know we don't use `submitEvent` directly anywhere yet.